Global styles: send theme object to setUserConfig

ramonjd opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Follow up to https://github.com/WordPress/gutenberg/pull/61271

Instead of destructuring object, just send it to the global styles provider callback via `setUserConfig`.

Props to @noisysocks for [catching it](https://github.com/WordPress/gutenberg/pull/61271#pullrequestreview-2064964677), and for giving me pause to discover how trivial the change is.

~~Should `setUserConfig` accept a function and an object?~~ 

Also, `setUserConfig` now accept a function and an object to update user global styles. So,

1. a callback function where the consumer needs the current config to merge changes (or whatever), and 
2. an object where the consumer just wants to overwrite the user config object completely

```js
const { setUserConfig } = useContext( GlobalStylesContext );

const newValue = {
    styles: { color: 'red' },
    settings: { test: 1 },
};

// With callback (merging)
setUserConfig( ( currentConfig ) => ( {
   styles: {
        ....currentConfig.styles,
        color: newValue.styles.color,
    },
    settings: newValue.settings,
} ) );

// Without (overwrite)
setUserConfig( newValue );
```

## Why?

The [global styles provider does the work](https://github.com/WordPress/gutenberg/blob/18e2d5c3f86eb2c00c5cb821d95957b45f0d9070/packages/editor/src/components/global-styles-provider/index.js#L84-L84) of splitting the properties into `styles`, `settings` and `_links`.


## Testing Instructions

There should be no regressions in the way global style variations or revisions work.

I'm using Twenty Twenty Four to test.

1. Fire up the site editor and open the global styles sidebar. Under "Browse styles", you'll find the theme variations. Apply a few theme variations and save each time.
2. Then open up the global styles revisions panel and apply some of your recently-created revisions.

There should be no JS errors, and the theme styles should apply to the canvas.

https://github.com/WordPress/gutenberg/assets/6458278/902a74da-53bf-41a9-87e3-bb898b3bd1dc

To test the `setUserConfig` arguments:

1. Switch and save various theme style variations
2. Fiddle with an individual block's styles and push them to global styles (in the advance panel of the block's settings)
3. Change global styles, then reset them. 

There should be no regressions.

ramonjd committed:
Instead of destructuring object, just send it. The global styles provider does the work.

noisysocks commented:
Thanks mate.

Missed one here—this could be simply `setCurrentlySelectedRevision( revision )`.

https://github.com/WordPress/gutenberg/blob/57b8b613d0507fd2c1ed0b441db09f76a43d579f/packages/edit-site/src/components/global-styles/screen-revisions/index.js#L94-L103

> Should setUserConfig accept a function and an object?

Yes 😀 this would be a nice cleanup. I think do it as part of these changes in this PR.

ramonjd commented:
> this would be a nice cleanup. I think do it as part of these changes in this PR.

Thanks for testing this one. Will do 👍🏻

ramonjd committed:
Remove unnecessary callback - use state function to set current revision

ramonjd committed:
Allow an object or a function in the setConfig function, which sets global styles.
Use the object arg for variations, pushing to global styles from block settings, and reset global styles.

noisysocks reviewed (approved):
Thanks! Much simpler, and this tests well for me.

Let's not backport to 6.6 as it's only code quality.

ramonjd commented:
Thanks for helping with this one. 🙇🏻

ellatrix commented:
I just cherry-picked this PR to the wp/6.6-beta-3 branch to get it included in the next release: 55abcaa92f
